### ISSUE TITLE
Can tests contain other tests, or provide input for other tests?

### ISSUE THREAD

**Issue by cmeeren** [haf/expecto#352]
Not sure this makes sense, but I'll try.

I am using Expecto for running end-to-end tests against an API. One test checks for the ability to create an API key. Other tests need an API key, and exercise endpoints using this key. There is also a test for deleting an API key.

Currently I have well-known API keys in the DB and simply use them for the tests that need an API key, but I'd like to avoid this. I'd like (if it's possible and makes sense) for the "can create API key" test to, if passed, supply the API key to the tests that need it, and for the "can delete API key" test to run at the end.

In other words, what I have here is a need for a common setup/teardown for a list of tests, but where the setup/teardown themselves are tests.

(I'd rather avoid using per-test setup/teardown since that would create a unique API key for every single test and thus triple the amount of requests.)

Is this possible? Does it make sense? I see from the readme that there's such a wealth of ways of defining/structuring/combining tests that it makes my head spin.

**Comment by haf** [comment]
```fsharp
let myTests (keyId: string) =
  testList [ ... ]
let main argv =
  use key = setupKey ()
  runTestsInAssemblyWithCLIArgs defaultConfig argv (myTests key.id)
```

**Comment by cmeeren** [comment]
Thanks, but is `setupKey` a test? Will it show in the test explorer and test results?

**Comment by haf** [comment]
There's really not anything "obvious" that solves your problem inside the test framework; but since you can program anything, you can choose your solution. My suggestion was that you put that logic outside of your running of the tests, so that `setupKey` sets up the key, and when the tests finish, the key gets deleted.

**Comment by cmeeren** [comment]
Thanks! That's what I wanted to know.

**Comment by haf** [comment]
@cmeeren Would it help to have an event system in Expecto? E.g. with "test starting", "test running", "test completed", "n/m tests completed", etc?

**Comment by cmeeren** [comment]
Well, it would be a powerful feature that would make it *possible* to create dependent tests (by setting mutable shared state), though I'm not in any way convinced it's a good idea to couple tests imperatively that way. (I just wanted to know if it was already supported in some manner.) Personally I'm having a hard enough time as it is wrapping my head around structuring tests in Expecto (and I'm currently the most F# proficient in our company, so I may still simply end up using another test framework to make it easier on everyone else.)

**Comment by cmeeren** [comment]
I have now understood `testFixture` for setup/teardown for each test, but is it possible to use it for common setup/teardown for multiple tests?

**Comment by haf** [comment]
Ok, I see what you're saying. So off the top of my head, perhaps something like:

```fsharp
let testListWith testListName disposableFactory (tests: ('thing -> Test) list) =
  testSequenced <| testList (sprintf "wrapper for %s" testListName) [
    let value = disposableFactory ()
    testList testListName (tests |> List.map (fun factory -> factory value))
    testCase (sprintf "%s disposing" testListName) <| fun () ->
      (value :> IDisposable).Dispose()
  ]
```

Usage:

```fsharp
let apiTestCase name testMethod =
  fun apiKey ->
    testCaseAsync name (testMethod apiKey)

[<Tests>]
let tests =
  testListWith "API Key" createAPIKey [
    apiTestCase "can fetch list" <| fun apiKey -> async {
      let! listing = API.FooBars.list apiKey [ Params.PageSize 10 ]
      listing |> Expect.isNonEmpty "Should return at least one item"
    }
  ]
```

**Comment by cmeeren** [comment]
Thanks a lot! The flexibility of Expecto is nice.

I would also like the actual tests using the API key (between setup/teardown) to run in parallel. I tried to place them in a test list under `testSequenced` and expected this inner list to run in parallel, but it seems it too is run in sequence. Is there any way to run a subset of `testSequenced` in parallel?

**Comment by joprice** [comment]
I was curious how to get this working for a setup function that requires `Task`, like a testcontainer or a local server, and got this variant working, where the setup function returns a task and another task to cleanup. Maybe someone will find it useful:

```fsharp
  let integrationTestFixtureList (name: string) (setup: unit -> Task<'b * (unit -> 'c)>) tests =
     testSequenced
      <| testList (sprintf "wrapper for %s" name) [
        let value = setup ()
        testList
          name
          (tests
           |> List.map (fun (name, factory) ->
             testCaseTask
               name
               (fun () -> task {
                 let! value, _ = value
                 do! factory value
               })
           ))
        testCaseTask (sprintf "%s disposing" name)
        <| fun () -> task {
          let! _, dispose = value
          do! dispose ()
        }
      ]
```
